heaps/min_heap.py

```diff
@@ -11,68 +11,100 @@ def __repr__(self):
         return str(self.value)
 
 
-
 class MinHeap:
 
     def __init__(self):
         self.store = []
 
-
-    def add(self, key, value = None):
+    def add(self, key, value=None):
         """ This method adds a HeapNode instance to the heap
             If value == None the new node's value should be set to key
-            Time Complexity: ?
-            Space Complexity: ?
+            Time Complexity: O(log n)
+            Space Complexity: O(1)
         """
-        pass
+        # if value is equal to none, set value to key
+        if value == None:
+            value = key
+        # Creates the HeapNode key and value pair
+        node = HeapNode(key, value)
+        self.store.append(node)  
+        self.heap_up(len(self.store) - 1)
 
     def remove(self):
         """ This method removes and returns an element from the heap
             maintaining the heap structure
-            Time Complexity: ?
-            Space Complexity: ?
+            Time Complexity: O(log n)
+            Space Complexity: O(1)
         """
-        pass
+        if len(self.store) == 0:
+            return None
+
+        # swaps the last leaf & the root node
+        self.swap(0, len(self.store) - 1)
+        # delete the last leaf (last item in the array)
+        min = self.store.pop()
+        # heap-down the new root, to reestablish the heap property
+        self.heap_down(0)
 
+        return min.value
 
-    
     def __str__(self):
         """ This method lets you print the heap, when you're testing your app.
         """
         if len(self.store) == 0:
-            return "[]"
+            return "[]"  # returns empty string
+        # otherwise returns the heap in an f string
         return f"[{', '.join([str(element) for element in self.store])}]"
 
-
     def empty(self):
         """ This method returns true if the heap is empty
-            Time complexity: ?
-            Space complexity: ?
+            Time complexity: O(1)
+            Space complexity: O(1)
         """
-        pass
-
+        return len(self.store) == 0
 
     def heap_up(self, index):
```

Comment:
👍  See above on space complexity, remember if it's recursive, it's likely to have space complexity.